### DISCUSSION TITLE
An indentation way to Lisp

### DISCUSSION THREAD

**Discussion by tiye** [Cirru/calcit-runner.nim#123] [Q&A]
Indentation based syntax, people may love it, people may hate it. I can't control that. My eyes prefer reading indentations, so I think it's a better tool for me. Lisp chose parentheses for expressing its syntax tree. I would say indentation based syntax is also able to do that.

The repo [calcit-runner](https://github.com/calcit-lang/calcit-runner) implemented a Clojure inspired scripting language, based on Cirru's rules of indentations. Here I got several examples...

Defining a function:

```cirru
defn f1 (x y)
  + x y
```

Creating a HashMap:

```cirru
{}
  :a 1
  :b $ [] 1 2 3 4
```

Binding local variables with `let`, notice the extra indentations for variables:

```cirru
let
    a 1
    b 2
  + a b
```

Conditional expression:

```cirru
cond
  (> a 1) (echo "|large than 1")
  (< a 1) (echo "|smaller than 1")
  true (echo "|just 1")
```

Indentation based syntax is powerful enough for expressing S-Expressions. Beginners may find it a lot easier to learn and to write as they first come to the world of Lisp, with far fewer parentheses.

There are also cases indentations may look clumsy, for example:

```cirru
echo
  + 1
    + 2
      + 3 4
```

Since this example is really simple, a quick solution can be:

```cirru
echo (+ 1 (+ 2 (+ 3 4)))
```

Well. It's a lot like lisp, just image there are more items in this example.

For such a situation, Cirru proposed `$` syntax, which [came from Haskell](typeclasses.com/featured/dollar).
The part after `$` is marked as an expression, so this example got clearer, with no parentheses:

```cirru
echo $ + 1 $ + 2 $ + 3 4
```

Another edge case is when lists and literals are mixed in an expression, indentation based syntax might be strange, for

```cirru
echo (+ 1 (+ 2 (+ 3 4))) :fin
```

one may want to write it with indentations like:

```cirru
echo
  + 1 (+ 2 (+ 3 4))
  :fin
```

however, this is not an identical structure to original one, it's actually:

```cirru
echo (+ 1 (+ 2 (+ 3 4))) (:fin)
```

you may find that `(:fin)` got an extra pair of parentheses, which is incorrect.
In order to make the structure useful, Cirru proposed another syntax for extracting a list, a comma `,` for extracting:

```cirru
echo (+ 1 (+ 2 (+ 3 4))) (, :fin)
```

When it's written with indentations, it looks like:

```cirru
echo
  + 1 (+ 1 (+ 3 4))
  , :fin
```

also be neater:

```cirru
echo
  + 1 $ + 2 $ + 3 4
  , :fin
```

In this way it's available to express lots of Lisp code.

One thing you would worry is Lisp has Macros. I am using Clojure, so I'm comparing it with Clojure:

```clojure
(defmacro m1 (x y)
  `(+ ~x ~y))
```

There is a backquote(`` ` ``) and several tilde(`~`) symbols. They are not transforming to indentations.

In calcit-runner, this macro can be written in this way:

```cirru
defmacro m1 (x y)
  quote-replace $ + (~ x) (~ y)
```

And for quoted data `'(1 2 3)`, it's written as:

```cirru
quote (1 2 3)
```

The result takes more key strokes, but it's fully available.

Well, there are more syntaxes in Clojure than I mentioned above, and even more in Racket or other dialects. I'm just saying indentation based syntax can express that, although might be less elegant.

Personally I believe Lisp with indentations is more friendly for beginners. People write code like in Python with a simple editor which is capable of maintain indentations. They don't have to learn EMACS for balancing parentheses. And also, who would say CoffeeScript has learning curves in its indentation based syntax?

**Comment by ksaj** [comment]


**Reply by tiye** [reply]
Well, Cirru is S-Expressions with indentations(and a few inline parentheses). Expressions are still lists. Try this demo http://repo.cirru.org/cirru-parser/ .

**Reply by tiye** [reply]
> I see nothing wrong with this. But it sure isn't lispy.

if by "lispy" you mean it's like Lisp, then no, it's not lispy.

> But the benefit of homoiconicity is that you can pretty much guess how something is supposed to go because that's how it goes absolutely everywhere else in the language.

to be honest I don't quite get you point. For "homoiconic", the [Wiki](https://en.wikipedia.org/wiki/Homoiconicity) says:

> A language is homoiconic if a program written in it can be manipulated as data using the language, and thus the program's internal representation can be inferred just by reading the program itself. For example, a Lisp program is written as a regular Lisp list, and can be manipulated by other Lisp code.[1] This property is often summarized by saying that the language treats "code as data".
>
> In a homoiconic language, the primary representation of programs is also a data structure in a primitive type of the language itself. This makes metaprogramming easier than in a language without this property: reflection in the language (examining the program's entities at runtime) depends on a single, homogeneous structure, and it does not have to handle several different structures that would appear in a complex syntax.

As for "code as data" in Cirru, inside the implementation in calcit-runner, a Clojure thread macro can be implemented like:

https://github.com/calcit-lang/calcit-runner/blob/master/src/calcit_runner/core_abstract.nim#L246-L254

```nim
  let codeThreadFirst = genCirru(
    ["defmacro", "->", ["base", "&", "xs"],
      ["if", ["empty?", "xs"], ["quote-replace", ["~", "base"]],
        ["&let", ["x0", ["first", "xs"]],
          ["if", ["list?", "x0"],
            ["recur", ["&concat", ["[]", ["first", "x0"], "base"], ["rest", "x0"]],
                      "&", ["rest", "xs"]],
            ["recur", ["[]", "x0", "base"], "&", ["rest", "xs"]]]]]]
  , coreNs)
```

which is identical to Cirru text format like:

```cirru
defmacro -> (base & xs)
  if (empty? xs) (quote-replace (~ base))
    &let (x0 (first xs))
      if (list? x0)
        recur (&concat ([] (first x0) base) (rest x0))
          , & (rest xs)
        recur ([] x0 base) & (rest xs)
```

sure it also can be translated to a Lisp directly(despite some difference in arguments spreading or recursion handling). So when you are talking about "homoiconicity", I would argue it's just "almost" Clojure. And Clojure is homoiconic.

> Some functions go on for pages. Cirru doesn't seem to allow you to space things out for clarity as a result of its lack of (lispy) homoiconicity.

Could you add an example for clarity?

It sounds to me like it was about differences about indentations and parentheses based S-Expressions. Indentations based syntax has its own pros and cons. I wouldn't just say indentations is superior.

**Reply by tiye** [reply]
> I prefer keeping consistent with the fact that every Lisp program is itself a list.

I would think it's a prejudice to consider "a list" as a syntax like `'(1 2 3)` with parentheses. Cirru merely turns some of the parentheses into indentations and looks difference. The data parsed from Cirru remains to be lists.

If you take a look at http://cirru.org/ , Cirru's higher goals is representing S-Expressions with a tree of DOM. That's going far beyond the concept of "list with parentheses".

**Reply by tiye** [reply]
well, English is not my daily language and I think we have to narrow down to aspects and better with real-world examples before we make this conversation too complex.

"lisp-like" might be a fuzzy word. I demonstrated the ability to describe macros with Cirru syntax so I said it's like Lisp. Meanwhile people may regard Lisp as just the parens syntax `(a (b c (d)))`. For me, like I just said, Cirru's goals are far beyond that.

> If you randomly took away parens...

I suppose you are criticizing Cirru before understanding the details of Cirru. Cirru is not randomly removing parens. If you take a piece of Cirru code, you will find that parens are balanced inside a single line. Indentations replaced parens that are using crossing lines in Lisp. Multi-lines code are hard to maintain by its own nature, it's hard in Cirru and it's hard in Lisp. The multi-lines problem needs to be solved with tools, and that's not currently discussed in this discussion post.

**Comment by ArneBab** [comment]
You can find another example of indentation-based lisp — implemented for Guile Scheme — at https://www.draketo.de/software/wisp

It’s homoiconic, generic, and backwards-compatible, and it provides an escape-hatch: If you decide that you actually prefer parentheses, or if someone asks you for a regular Scheme version, you can use the shipped wisp2lisp-script to turn your files into parentheses-based code.

**Reply by tiye** [reply]
at first I thought it was https://github.com/wisp-lang/wisp but turned out they are totally 2 projects. interesting.

well, Calcit is special since itself was first targeting syntax tree editor like http://cirru.org/ as the goal, and the indentation syntax is a compatible solution to that tree editor. so I don't have plan for parsing another Lisp syntax in this project.

meanwhile I do agree that in many scenarios people would love the code wrapped with parentheses. so in the format the code  in Calcit printed, it sometimes wraps with parentheses like a Lisp. and in APIs docs http://apis.calcit-lang.org/ there's also a Lisp style for displaying.
